pdfPreviewInMsg/background.js

```diff
@@ -0,0 +1,166 @@
+/*
+----------------------------------------------------------------
+This extension for Thunderbird was made by : e-gaulue & rholeczy
+                        © 2021
+----------------------------------------------------------------
+*/
+var initialized;
+
+
+/**
+ * Preview a pdf in the mail.
+ * 
+ * @param {*} pdf 
+ * @param {*} canvas 
+ * @param {*} nbP 
+ * @param {*} tabid 
+ * @param {*} filename 
+ * @param {*} messageid 
+ */
+async function pdfPreviewRH(pdf, canvas, nbP, tabid, filename, messageid) {
+  let page = await pdf.getPage(nbP);
+  let viewport = await page.getViewport({ scale: 1.0 });
+  let scale = Math.min(
+    canvas.width / viewport.width,
+    canvas.height / viewport.height
+  );
+  let thumbViewPort = await page.getViewport({ scale });
+  let context = canvas.getContext("2d");
+  let task = page.render({ canvasContext: context, viewport: thumbViewPort });
+  task.promise.then(() => {
+    // Convert the pdf's page to an image.
+    let urlImage = canvas.toDataURL("image/jpeg"); 
+    // Send a message to the tab to do ???
+    browser.tabs.sendMessage(tabid, {
+      type: "handle_me",
+      image: urlImage,
+      part_name: filename,
+      numPage: nbP,
+      messageId: messageid,
+    });
+  });
+}
+
+
+/**
+ * <Description>
+ * 
+ * @param {*} tabid 
+ * @param {*} messageid 
+ * @returns 
+ */
+async function execWhenInitialized(tabid, messageid) {
+  if (initialized === false) {
+    setTimeout(execWhenInitialized, 50, tabid, messageid);
+    return;
+  }
+
```

Comment:
`initialized` is set to true, when your display script has send the message, correct? So could we not call `execWhenInitialized()`, when that happens? In the `onMessage` Listener, instead of here and doing a timeout loop?

Reply:
Well I believe it's not necessary anymore. At the beginning, we used to have:
```
browser.messageDisplayScripts.register({ js: [{ file: "pdf.js" },{ file: "pdf.worker.js" },{ file: "initial.js" }] });
```
It was really long and we needed to be sure everything was ready in the message pane before to send the attachments from the background. Then you gave me the idea to do all in the background and then just send the resulting images through message in the mail pane. As it's just thumbs, those messages should stayed small by the way.

So presently, I think we just don't need to wait anymore. `initial.js` just add a listener while the background script need to get the attachments of the message. Of course those two events execute concurrently, but I can't imagine the listener is not ready yet when we send the images from the background.

So I think we can remove all this stuff.

Reply:
I would keep the logic, that the initial.js scripts "request" the information from the background. That is the most clean approach. I think you do not need the onMessageDisplayed listener, as that is effectively covered by the loaded display script.

Reply:
Romain has just test without and it doesn't work.

If we execute `execWhenInitialized` when it happens, we need to get `tabid` and 'messageid`. I don't think displayscripts can get those. But maybe we can get it from `background.js, in this case we needn't `onDisplayedMessage` anymore.

We can give it a try,

Reply:
the onMessage event has a "sender" parameter, which has the tab of the sender, and the [messageDisplay API](https://webextension-api.thunderbird.net/en/78/messageDisplay.html#getdisplayedmessage-tabid) gets you the displayed message in that tab.

Reply:
It's done jobisoft ;) I did exactly what you think you can take a look, i already push it

Reply:
The addon developer matrix chat:
https://developer.thunderbird.net/add-ons/community

I am "TbSync" there, and online during europe working hours.